[JSC] Micro-optimize String equal operation with UChar / LChar

Constellation opened the pull request:
#### bacdbdaaf1824b08dc7ad35035d5d8358be623c5
<pre>
[JSC] Micro-optimize String equal operation with UChar / LChar
<a href="https://bugs.webkit.org/show_bug.cgi?id=268684">https://bugs.webkit.org/show_bug.cgi?id=268684</a>
<a href="https://rdar.apple.com/122224476">rdar://122224476</a>

Reviewed by Ryosuke Niwa.

This patch micro-optimizes String equal operation with different characters (UChar* and LChar*).

                                                         ToT                     Patched

    todomvc-javascript-es5-json-parse              37.6466+-0.1862     ^     37.1991+-0.1560        ^ definitely 1.0120x faster
    todomvc-javascript-es6-webpack-json-parse
                                                   58.9239+-0.3310     ^     58.2251+-0.1931        ^ definitely 1.0120x faster

* Source/WTF/wtf/text/StringCommon.h:

Canonical link: <a href="https://commits.webkit.org/274064@main">https://commits.webkit.org/274064@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1a508d6c2912697367573948436dd1cf8576feb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37715 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16608 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39954 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40255 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33555 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19253 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13804 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31918 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38282 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13967 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33017 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12212 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12136 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41514 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/31447 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34099 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34137 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38035 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/37420 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12735 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10274 "Found 1 new test failure: ipc/message-listener-async-message-reply-id.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36207 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14150 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/44327 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13121 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9060 "Found 157 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout, jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout-dfg-eager-no-cjit, jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout-no-cjit, jsc-layout-tests.yaml/js/script-tests/regress-150580.js.layout-no-llint, stress/array-flatmap.js.dfg-eager-no-cjit-validate, wasm.yaml/wasm/gc/any.js.default-wasm, wasm.yaml/wasm/gc/any.js.wasm-bbq, wasm.yaml/wasm/gc/any.js.wasm-collect-continuously, wasm.yaml/wasm/gc/any.js.wasm-eager, wasm.yaml/wasm/gc/any.js.wasm-eager-jettison ... (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4897 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13463 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/WTF/wtf/text/StringCommon.h:
```diff
@@ -344,13 +344,36 @@ ALWAYS_INLINE bool equal(const LChar* a, const UChar* b, unsigned length)
     }
-    // Otherwise, we just do a naive loop.
-#endif
+    if (length >= 4) {
```
Since the cases < 8 don't use SIMD intrinsics, couldn't these also be enabled for x64 without further changes?

Constellation replied:
The code is relying on that we have `length >= 8` loop, which x64 does not have.

Constellation committed:
[JSC] Micro-optimize String equal operation with UChar / LChar
https://bugs.webkit.org/show_bug.cgi?id=268684
rdar://122224476

Reviewed by Ryosuke Niwa.

This patch micro-optimizes String equal operation with different characters (UChar* and LChar*).

                                                         ToT                     Patched

    todomvc-javascript-es5-json-parse              37.6466+-0.1862     ^     37.1991+-0.1560        ^ definitely 1.0120x faster
    todomvc-javascript-es6-webpack-json-parse
                                                   58.9239+-0.3310     ^     58.2251+-0.1931        ^ definitely 1.0120x faster

* Source/WTF/wtf/text/StringCommon.h:

Canonical link: https://commits.webkit.org/274064@main
